### PR TITLE
storage: improve SSH tunnel failure errors

### PR DESCRIPTION
**Pull request by benesch**
We've been seing errors like

    failed to close ssh tunnel: the master connection failed

in production. The actual *reason* that the SSH tunnel failed is buried inside the returned error type, and not part of the `fmt::Display` representation.

This commit adjust the logging to route through `anyhow::Error`, which can be convinced to print all underlying error causes via the `{:#}` format specifier.

### Motivation

  * This PR improves internal debuggability of SSH tunnels.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a (only helps internal debugging)

### CODE REVIEW

File: src/storage-client/src/types/connections.rs [hunk 1]
```diff
@@ -1016,11 +1017,11 @@ impl SshTunnel {
...
         mz_ore::task::spawn(|| format!("{}_ssh_session", debug_str), async move {
             let _: Result<(), _> = rx.await;
-            session
-                .close()
-                .await
-                .err()
-                .map(|e| tracing::error!("failed to close ssh tunnel: {e}"));
+            if let Err(e) = session.close().await {
+                // Convert to `anyhow::Error` to include error source, which can
+                // contain key details about the nature of the failure.
```
Comment (guswynn) [inline src/storage-client/src/types/connections.rs:1022]:
```suggestion
                // Convert to `anyhow::Error` whose `Debug` impl automatically
                // consumes the entire chain of error `source`s, which the std
                // error trait doesn't yet provide.
```

File: src/postgres-util/src/lib.rs [hunk 2]
```diff
@@ -396,15 +398,15 @@ impl Config {
...
-                        .await
-                        .err()
-                        .map(|e| tracing::error!("failed to close ssh tunnel: {e}"));
+                    if let Err(e) = connection.await {
+                        warn!("postgres connection failed: {e}");
+                    }
+                    if let Err(e) = session.close().await {
+                        // Convert to `anyhow::Error` to include error source,
+                        // which can contain key details about the nature of the
+                        // failure.
```
Comment (guswynn) [inline src/postgres-util/src/lib.rs:407]: same as below

**Review by guswynn — APPROVED** [review]
looks good, but I have a wording accuracy nit

File: src/storage-client/src/types/connections.rs (hunk 1 above)
Comment (benesch) [reply src/storage-client/src/types/connections.rs:1022]: Do you mean `Display` impl?

**Review by guswynn — APPROVED** [review]

**Review by benesch — COMMENTED** [review]
Updated the comment per your feedback and marked for automerge. Thanks, @guswynn!

**Pull request merged**